circuit/src/native.rs

```diff
@@ -0,0 +1,561 @@
+use std::collections::HashMap;
+
+use crate::eddsa::native::{PublicKey, Signature};
+use halo2::{
+	arithmetic::Field,
+	halo2curves::{bn256::Fr, FieldExt},
+};
+
+const NUM_NEIGHBOURS: usize = 6;
+const NUM_ITERATIONS: usize = 20;
+const INITIAL_SCORE: u128 = 1000;
+
+#[derive(Debug, Clone)]
+struct Opinion {
+	sig: Signature,
+	message_hash: Fr,
+	scores: [(PublicKey, Fr); NUM_NEIGHBOURS],
+}
+
+impl Opinion {
+	pub fn new(
+		sig: Signature, message_hash: Fr, scores: [(PublicKey, Fr); NUM_NEIGHBOURS],
+	) -> Self {
+		Self { sig, message_hash, scores }
+	}
+}
+
+impl Default for Opinion {
+	fn default() -> Self {
+		let sig = Signature::new(Fr::zero(), Fr::zero(), Fr::zero());
+		let message_hash = Fr::zero();
+		let scores = [(PublicKey::default(), Fr::zero()); NUM_NEIGHBOURS];
+		Self { sig, message_hash, scores }
+	}
+}
+
+struct EigenTrustSet {
+	set: [(PublicKey, Fr); NUM_NEIGHBOURS],
+	ops: HashMap<PublicKey, Opinion>,
+}
+
+impl EigenTrustSet {
+	pub fn new() -> Self {
+		Self { set: [(PublicKey::default(), Fr::zero()); NUM_NEIGHBOURS], ops: HashMap::new() }
+	}
+
+	pub fn add_member(&mut self, pk: PublicKey) {
+		let pos = self.set.iter().position(|&(x, _)| x == pk);
+		// Make sure not already in the set
+		assert!(pos.is_none());
+
+		let first_available = self.set.iter().position(|&(x, _)| x == PublicKey::default());
+		let index = first_available.unwrap();
+
+		// Initial score for new member is zero.
+		self.set[index] = (pk, Fr::zero());
+	}
+
+	pub fn remove_member(&mut self, pk: PublicKey) {
+		let pos = self.set.iter().position(|&(x, _)| x == pk);
+		// Make sure already in the set
+		assert!(pos.is_some());
+
+		let index = pos.unwrap();
+		self.set[index] = (PublicKey::default(), Fr::zero());
+
+		self.ops.remove(&pk);
+	}
+
+	pub fn update_op(&mut self, from: PublicKey, op: Opinion) {
+		let pos_from = self.set.iter().position(|&(x, _)| x == from);
+		assert!(pos_from.is_some());
+
+		// Initial score is updated only when the opinion is given/signed.
+		let initial_score = Fr::from_u128(INITIAL_SCORE);
+		let index = pos_from.unwrap();
+		self.set[index] = (from, initial_score);
+
+		self.ops.insert(from, op);
+	}
+
+	pub fn converge(&mut self) -> [Fr; NUM_NEIGHBOURS] {
+		let mut filtered_set = self.set.clone();
+		let mut filtered_ops: HashMap<PublicKey, Opinion> = HashMap::new();
```

Comment:
The goal is to avoid mutating the original variables, so here we should do `self.ops.clone()`